File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys
from collections import deque

N = int(sys.stdin.readline().rstrip())
A = list(map(int,sys.stdin.readline().split()))
plus, minus, mul, div = map(int,sys.stdin.readline().split())

queue = deque()
queue.append([A[0],plus,minus,mul,div,1])

max_res = -10000000000
min_res = 10000000000

while queue:
    num, plus, minus, mul, div, depth = queue.popleft()

    if depth == N:
        max_res = max(max_res,num)
        min_res = min(min_res,num)
        continue

    next_num = A[depth]

    if plus > 0:
        queue.append([num+next_num, plus-1,minus,mul,div,depth+1])
    if minus > 0:
        queue.append([num-next_num,plus,minus-1,mul,div,depth+1])
    if mul > 0:
        queue.append([num * next_num, plus, minus, mul-1, div, depth + 1])
    if div > 0:
        if num < 0 and next_num > 0:
            change_num = -(abs(num) // next_num)
            queue.append([change_num, plus, minus, mul, div-1, depth + 1])
        else:
            queue.append([num // next_num, plus, minus, mul, div-1, depth + 1])

print(max_res)
print(min_res)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.util.*;
import java.io.*;

class Main {
  static int N;
  static int[] numbers;
  static int max = Integer.MIN_VALUE;
  static int min = Integer.MAX_VALUE;
  public static void main(String[] args) throws IOException {
    BufferedReader br  = new BufferedReader(new InputStreamReader(System.in));

    N = Integer.parseInt(br.readLine());

    String[] st = br.readLine().split(" ");
    
    numbers = new int[N];
    for (int i = 0; i < N; i++) numbers[i] = Integer.parseInt(st[i]);

    String[] func = br.readLine().split(" ");

    int add = Integer.parseInt(func[0]); 
    int minus = Integer.parseInt(func[1]); 
    int multi = Integer.parseInt(func[2]); 
    int div = Integer.parseInt(func[3]);

    dfs(1 ,numbers[0], add, minus, multi, div);
    System.out.println(max);
    System.out.println(min);

  }
  static void dfs(int idx, int num,int add, int minus, int multi, int div) {
    if (idx == N) {
      max = Math.max(num, max);
      min = Math.min(num, min);
      return;
    }

    if (add > 0) {
      dfs(idx+1, num + numbers[idx], add-1, minus, multi, div);
    }
    if (minus > 0) {
      dfs(idx+1, num - numbers[idx], add, minus-1, multi, div);
    }
    if (multi > 0) {
      dfs(idx+1, num * numbers[idx], add, minus, multi-1, div);
    }
    if (div > 0) {
      int next;
      if (num < 0) {
        next = -(-num / numbers[idx]);
      } else {
        next = num / numbers[idx];
      }
      dfs(idx+1, next, add, minus, multi, div-1);
    }
  }
} 